IN REVIEW: Dynamic Node Groups Proposal (#8593)
gardener/gardener: docs/proposals/24-dynamic-node-groups-for-shoot-cluster-autoscaler.md

          - [Pool-0:2:1:1:2 (min:max:maxsurge:maxunavail:numzones)](#pool-02112-minmaxmaxsurgemaxunavailnumzones)
          - [Scan-0](#scan-0-2)
          - [Scan-1](#scan-1-2)
    - [4. Shoot Spec Enhancement](#4-shoot-spec-enhancement)
    - [TODO](#todo)

## Summary

Presently, the `NodeGroups` for the shoot `cluster-autoscaler`  are given static `Minimum` and `Maximum` sizes computed in the gardener extension providers during shoot reconciliation. This has several disadvantages since the static computation preferences zones that are specified earlier over zones specified later. In addition, for cases where the number of zones exceeds the `Max` specified for the worker pool, the node-group is assigned a zero `Max` size.

This document outlines deficiences in the existing solution and proposes moving to the CA's node group auto-discovery mechanism where node groups can be discovered and sized at runtime. 

## Motivation

TODO: Add Sequence Diagram for the below.

Let us first study how `MachineDeployments` are currently generated and how node groups are currently computed for the `cluster-autoscaler`.

### Existing Solution

#### Worker Actuator Reconciliation 

> The Worker [generic.Actuator.Reconcile](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go#L42) implementation delegates to [genericactuator.GenerateMachineDeployments](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/genericactuator/interface.go#L46) to generate [worker.MachineDeployments](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/machines.go#L39) for a shoot clulster during shoot reconciliation.

Reviewer: ```suggestion
The Worker [generic.Actuator.Reconcile](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go#L42) implementation delegates to [genericactuator.GenerateMachineDeployments](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/genericactuator/interface.go#L46) to generate [worker.MachineDeployments](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/machines.go#L39) for a shoot cluster during shoot reconciliation.
```

> These worker MCD's are converted into MCM [MachineDeployments](https://pkg.go.dev/github.com/gardener/machine-controller-manager@v0.50.0/pkg/apis/machine/v1alpha1#MachineDeployment)'s and then deployed into the shoot control plane.

Reviewer 2: What does `MCD` mean?

Author: It means machine deployments. I will change to use the full form

Reviewer 2: Thanks.

> These Worker MachineDeployments's are converted into MCM [MachineDeployments](https://pkg.go.dev/github.com/gardener/machine-controller-manager@v0.50.0/pkg/apis/machine/v1alpha1#MachineDeployment)'s and then deployed into the shoot control plane.

Reviewer: Maybe you can also include the link where this is done.
https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go#L204 is the method where this is done. This only helps the reviewer to quickly navigate to the code where this is currently done.

> These worker deployments are also set on the [Worker Status](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go#L470)

Reviewer 2: ```suggestion
These worker deployments are also set on the [Worker Status](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go#L470).
```

The [Cluster Autoscaler Deployer](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/pkg/component/clusterautoscaler/cluster_autoscaler.go#L103) then constructs the deployment command string via [computeCommand](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/pkg/component/clusterautoscaler/cluster_autoscaler.go#L397) to iterate through these worker deployments and create static node groups as follows:

```go
	for _, machineDeployment := range c.machineDeployments {
		command = append(command, fmt.Sprintf("--nodes=%d:%d:%s.%s", machineDeployment.Minimum, machineDeployment.Maximum, c.namespace, machineDeployment.Name))
	}
```
This is the format `--nodes=nodeGroupMin:nodeGroupMax:nodeGroupName`

> Effectively we currently have a `1:1` correspondence between a `MachineDeployment` and a CA `NodeGroup`. The `NodeGroup` is the abstraction used by the CA to represent a set of nodes that have the same capacity and set of labels and within which scale-up and scale-down oeprations cna be performed.

Reviewer 2: ```suggestion
Effectively we currently have a `1:1` correspondence between a `MachineDeployment` and a CA `NodeGroup`. The `NodeGroup` is the abstraction used by the CA to represent a set of nodes that have the same capacity and set of labels and within which scale-up and scale-down operations can be performed.
```

#### MachineDeployment Creation Details

> Most garden extension providers have a helper method [generateMachineConfig](https://github.com/gardener/gardener-extension-provider-gcp/blob/c3439f0a7e6bd76dda464be529396e0590983c55/pkg/controller/worker/machines.go#L87) that iterates over the worker pools and the pool zones and generates the [worker.MachineDeployment]'s for each worker pool and zone combination as can be seen from the snippet below.

Reviewer 2: Why the brackets? Did you miss to add a link?

Author: yes. will correct.

```go
for _, pool := range w.worker.Spec.Pools {

> 		zoneLen := int32(len(pool.Zones))

Reviewer 2: Please fix the indentation.

Reviewer 2: Thanks for fixing this.
Please check the line below, which is not indented at all.

Author: Hi scheerer, apologies for delay, can you let me know which line is not indented ? The block for `DistributeOverZones` is sadly 8 spaces indented by the markdown viewer for god-knows what reason. (There is actually only one space in reality)

###### Pool-0:1:1:1:2 (min:max:maxsurge:maxunavail:numzones)
```
MachineDeployment_Z0(Minimum=0|Maximum:1|MaxSurge:1|MaxUnavailable:1)
MachineDeployment_Z1(Minimum=0|Maximum:0|MaxSurge:0|MaxUnavailable:0)
```
Consider the case above where `numzones:2` > worker pool `max:1`, We can see that the machinedeployment for `Z1` is useless as its `Min`, `Max`, `MaxSurge` and `MaxUnavailable` are all zero making the corresponding CA `NodeGroup` also useless.

###### Pool-3:5:1:1:2 (min:max:maxsurge:maxunavail:numzones)
```
MachineDeployment_Z0(Minimum=2|Maximum:3|MaxSurge:1|MaxUnavailable:1)
MachineDeployment_Z1(Minimum=1|Maximum:2|MaxSurge:0|MaxUnavailable:0)
```
Consider the case above - `MaxSurge`  for `Z1` is computed as `0`

The above cases are not un-common. Thre are `~65` Worker Pools in LIVE Landscape with `0 < WorkerPoolMax < WorkerPoolZones`

### Deficiencies of Existing Solution

1. The current static node group distribution skews distribution towards earlier specified zones over later specified zones.
2. If zone length exceeds worker pool max, then later node groups get zero values effectively becoming useless.

> 3. If the earlier node groups have quota issues, then later node groups have reduced capacity and cannot meet the specified availability. 

Reviewer 2: Is it really about quota issues or just generally about instance capacity issues? I suspect it is rather the later, i.e. the infrastructure has problems providing new instances in one availability zone.

Author: changing from quota to capacity. it corresponds to CA `cloudprovider.OutOfResourcesErrorClass` inside `InstanceStatus.ErrorInfo.ErrorClass`

### Goals

1. We remove the current static node group assignment and move to a dynamic node group policy leveraging [node-group-auto-discovery](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/main.go#L156)
2. We change our MCM [CloudProvider](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go#L52) implementaion to support dynamic node groups and node group auto-discovery.
3. We change the deployment of the CA Deployer to construct the command string differently.

> 4. We fix invalid shoot clusters on all landscapes where `WorkerPool.Maxium >= WorkerPool.NumZones` 

Reviewer 2: Is the comparison sign correct? I would rather expect `Maximum < Zones` to be the issue.

Author: oops

> 4. We fix invalid shoot clusters on all landscapes where `WorkerPool.Maxium < WorkerPool.NumZones` 

Reviewer: ```suggestion
4. We fix invalid shoot clusters on all landscapes where `WorkerPool.Maximum < WorkerPool.NumZones` 
```

   1. We also introduce validation for new shoot clusters to ensure this constraint is honored.

### Non-Goals

1. No revamp of core `cluster-autoscaler` mechanics is considered in this proposal.

## Proposal

### 0. Add PoolSize Validation / Fix Existing Clusters 

> As was described earlier, our current distribution logic for worker pools can result in creating non-sensical NodeGroups where the `MaxSize` of the NodeGroup is computed as `0`. Since we will preserve the current distribution logic in our `Backward-Compatible` NodeGroup sizing strateg, we should add a validation check to reject shoot specs where `WorkerPool.Maximum > WorkerPool.NumZones`. We should also fix existing shoot specs.

Reviewer 2: ```suggestion
As was described earlier, our current distribution logic for worker pools can result in creating non-sensical NodeGroups where the `MaxSize` of the NodeGroup is computed as `0`. Since we will preserve the current distribution logic in our `Backward-Compatible` NodeGroup sizing strategy, we should add a validation check to reject shoot specs where `WorkerPool.Maximum > WorkerPool.NumZones`. We should also fix existing shoot specs.
```

Again the comparison seems wrong.

Author: oops.fixing

> As was described earlier, our current distribution logic for worker pools can result in creating non-sensical NodeGroups where the `MaxSize` of the NodeGroup is computed as `0`. Since we will preserve the current distribution logic in our `Backward-Compatible` NodeGroup sizing strateg, we should add a validation check to reject shoot specs where `WorkerPool.Maximum < WorkerPool.NumZones`. We should also fix existing shoot specs.

Reviewer: ```suggestion
As was described earlier, our current distribution logic for worker pools can result in creating non-sensical NodeGroups where the `MaxSize` of the NodeGroup is computed as `0`. Since we will preserve the current distribution logic in our `Backward-Compatible` NodeGroup sizing strategy, we should add a validation check to reject shoot specs where `WorkerPool.Maximum < WorkerPool.NumZones`. We should also fix existing shoot specs.
```

The above will permit the MCM CloudProvider to discover all the Machine Deployments for the shoot cluster.

### 3. Dynamic Node Group Sizing.

The [cloudprovider.CloudProvider](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/cloudprovider/cloud_provider.go#L100) interface is the facade between the CA and the cloud platform.

[cloudprovider.NodeGroup](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/cloudprovider/cloud_provider.go#L163) is the CA abstraction used to represent the set of cloud-provider nodes which are in-turn represented by [cloudprovider.Instances](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/cloudprovider/cloud_provider.go#L238)

> The CA carries out scale-down and scale-up operations within a `NodeGroups` respecting the constraints given by the `NodeGroup.MinSize()` and `NodeGroup.MaxSize()` methods.

Reviewer: ```suggestion
The CA carries out scale-down and scale-up operations within a `NodeGroup` respecting the constraints given by the `NodeGroup.MinSize()` and `NodeGroup.MaxSize()` methods.
```

> The CA runs a reconcile loop every `scanPeriod` interval (default: `10s`). At the beginningo of the reconcile loop, the CA invokes `CloudProvider.Refresh()` to permit the cloud provider implementation to update its cache and then issues a call to `CloudProvider.NodeGroups()` to retrieve the latest `[]NodeGroup`.  The MCM CloudProvider implements this by interrogating the machine deployments.

Reviewer 2: ```suggestion
The CA runs a reconcile loop every `scanPeriod` interval (default: `10s`). At the beginning of the reconcile loop, the CA invokes `CloudProvider.Refresh()` to permit the cloud provider implementation to update its cache and then issues a call to `CloudProvider.NodeGroups()` to retrieve the latest `[]NodeGroup`.  The MCM CloudProvider implements this by interrogating the machine deployments.
```

Each `NodeGroup` continues to be associated with its corresponding `MachineDeployment` for the zone.

We will offer 2 strategies for node-group sizing: *Lax Greedy Sizing* and *Equitable Sizing*

### 3.1 Lax-Greedy Sizing
1. If this strategy is selected, then for each `NodeGroup` the `NodeGroup.MaxSize()` is initially returned as the `MachineDeployment.Spec.PoolMaximum` and the `NodeGroup.MinSize()` is given as the `0` if there are no nodes provisioned for the shoot cluster.
1. The CA will execute scale-up/scale-down activities for the node groups if needed. If a NodeGroup does not have quota it will back-off and try another NodeGroup.

> 1. For the next scane, we compute `NodeGroup.MaxSize()` as follows

Reviewer 2: ```suggestion
1. For the next scan, we compute `NodeGroup.MaxSize()` as follows
```

>      1. `NodeGroup.MaxSize()=      PoolMaximum-CountOfAllNodesMaterializedInOtherNodeGroups)`

Reviewer 2: ```suggestion
     1. `NodeGroup.MaxSize() = PoolMaximum-CountOfAllNodesMaterializedInOtherNodeGroups)`
```

#### Pros/Cons
1. Pro: This strategy has the primary advantage that workload can be provisioned within the zone for the NodeGroup. It can be considered as work-load friendly
2. Pro: It can handle back-off well. If nodes can't be provisioned in one node group, the other node group can take up the slack.

> 3. Con: It isn't backward compatabile with our current distribution logic.

Reviewer 2: ```suggestion
3. Con: It isn't backward compatible with our current distribution logic.
```

> 4. Con: We need to add support for scale from zero for extended/ephmeral resources. We have [Gardener CA #132](https://github.com/gardener/autoscaler/issues/132) for this.

Reviewer 2: ```suggestion
4. Con: We need to add support for scale from zero for extended/ephemeral resources. We have [Gardener CA #132](https://github.com/gardener/autoscaler/issues/132) for this.
```

> 

Reviewer 2: Another downside of this is that it may not use zones at all, i.e. all instances may stay in one availability zone even if multiple are configured. While this may be advantageous short-term, it may impact availability if the zone goes down. The potential impact in terms of availability/balance needs to be clear to users.

Author: adding as Con

Reviewer: >Another downside of this is that it may not use zones at all, i.e. all instances may stay in one availability zone even if multiple are configured.  

Ideally this should not be the concern of CA on which node group it should scale up. Instead the onus should be on the consumer to appropriately set TSC or alternative policies to force distribution of pods across zones which in turn will force CA to choose another zone, if a new node in another zone cannot be launched then the pod will remain in pending state (unschedulable)

Reviewer: @himanshu-kun pointed that CA also provides [balance-similar-node-groups](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/main.go#L181) CLI flag which allows CA to make best effort to balance launching of nodes amongst similar node groups. AFAIK this would also help distribute the nodes amongst similar node groups and should continue to work even with the `Lax` strategy. However we should probably test this.

Reviewer 2: Agreed. Users need to ensure that their workload has topology spread constraints set.

2. Con: Unfortunately, it doesn't handle back-off well.

#### Examples

##### Pool-3:4:2:2:3 (min:max:maxsurge:maxunavail:numzones)

###### Scan-0
```
NG-0: Min:1, Max:2, Launched: 0
NG-1: Min:1, Max:1, Launched: 0
NG-2: Min:1, Max:1, Launched: 0
```

###### Scan-1

```
NG-0: Min:1, Max:2, Launched: 1
NG-1: Min:1, Max:1, Launched: 1
NG-2: Min:1, Max:1, Launched: 0
```

###### Scan-2

```
NG-0: Min:1, Max:2, Launched: 2
NG-1: Min:1, Max:1, Launched: 1
NG-2: Min:1, Max:1, Launched: 1
```

###### Pool-0:2:1:1:2 (min:max:maxsurge:maxunavail:numzones)

* PoolMax=2
* Let us take the case where `NG-0` is out of quota

###### Scan-0
```
NG-0: Min:0, Max:1, Launched: 0
NG-1: Min:0, Max:1, Launched: 0
```

###### Scan-1
```
NG-0: Min:0, Max:1, Launched: 0 (backoff)
NG-1: Min:0, Max:1, Launched: 1
```
* No further progress since `NG-1` has reached its computed `Max` and no nodes can be provisioned in `NG-0`.  This is our current behaviour unfortunately.

### 4. Shoot Spec Enhancement

We add a `nodeGroupSizingStrategy` in our the `clusterAutoscaler` section of our `shoot` YAML

```yaml
clusterAutoscaler:
  sizing: Lax-Greedy|Backward-Compatible
```

### TODO

> 1. Discuss `MaxUnavailable` and `MaxSurge` handling. SHould this be copied or distributed ?

Reviewer 2: ```suggestion
1. Discuss `MaxUnavailable` and `MaxSurge` handling. Should this be copied or distributed ?
```
